### PR TITLE
👩‍💼 Users endpoints migration to node

### PR DESCRIPTION
**Pull request by romanesko**
## Description
added endpoints:
/users/:id
/users
regarding https://www.notion.so/twake/Rest-APIs-07c0294728794901b8c8a4518ad5852c

## Related Issue
https://app.zenhub.com/workspaces/twake-5f06c5bf7796310019354875/issues/linagora/twake/1167

## How Has This Been Tested?
added e2e tests in twake/backend/node/test/e2e/users.spec.ts 
tested on local database (casandra only tested)

## Types of changes
- [    ] Bug fix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [    ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ x ] My code follows the code style of this project.
- [ x ] My change requires a change to the documentation.
- [    ] I have updated the documentation accordingly.
- [ x ] I have read the **CONTRIBUTING** document.
- [    ] I have added the Signed-off-by statement at the end of my commit message.

### CODE REVIEW

File: twake/backend/node/jest.config.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// jest.config.js
+// Sync object
+/** @type {import('@jest/types').Config.InitialOptions} */
+const config = {
+  verbose: true,
+  rootDir: ".",
+  preset: "ts-jest"
+};
+
+module.exports = config;
```
Comment (RomaricMourgues) [inline twake/backend/node/jest.config.js:10]: Why is this file needed ? (just to know)

File: twake/backend/node/test/e2e/users.spec.ts [hunk 2]
```diff
@@ -40,15 +43,109 @@ describe.skip("The /users API", () => {
...
-        id,
+        error: "Not Found",
+        message: `User ${id} not found`,
+        statusCode: 404,
+      });
+      done();
+    });
+
+    it("should 200 and big response for myself", async done => {
+      const id = "9ba49092-af64-11eb-90a9-0242ac150004";
```
Comment (RomaricMourgues) [inline twake/backend/node/test/e2e/users.spec.ts:56]: I suppose keeping this hard coded will break the ci tests ?

File: twake/backend/node/jest.config.js (hunk 1 above)
Comment (romanesko) [reply twake/backend/node/jest.config.js:10]: For requires by JetBrains Idea for running tests in debug mode (very convenient for TDD)

File: twake/backend/node/test/e2e/users.spec.ts (hunk 2 above)
Comment (romanesko) [reply twake/backend/node/test/e2e/users.spec.ts:56]: Yep, it is…   gonna fix this

File: twake/backend/node/jest.config.js (hunk 1 above)
Comment (romanesko) [reply twake/backend/node/jest.config.js:10]: I'll add it to .gitignore better

File: twake/backend/node/src/services/user/services/companies/service.ts [hunk 3]
```diff
@@ -89,4 +107,29 @@ export class CompanyService implements CompaniesServiceAPI {
...
+    role: "guest" | "admin" | "member",
+  ): Promise<void> {
+    const entity = await this.companyUserRepository.findOne({
+      group_id: companyPk.id,
+      user_id: userPk.id,
+    });
+    console.log({
+      group_id: companyPk.id,
+      user_id: userPk.id,
+    });
```
Comment (RomaricMourgues) [inline twake/backend/node/src/services/user/services/companies/service.ts:129]: Console log to remove

File: twake/backend/node/src/services/user/web/controller.ts [hunk 4]
```diff
@@ -6,40 +6,200 @@ import { ResourceListResponse } from "../../../services/types";
...
     if (!user) {
       reply.notFound(`User ${request.params.id} not found`);
 
       return;
     }
 
+    // const getCompaniesIds = (companyId: string) =>
+    //   this.service
+    //     .getUserCompanies({ id: companyId })
+    //     .then(a => a.getEntities().map(a => a.group_id));
```
Comment (RomaricMourgues) [inline twake/backend/node/src/services/user/web/controller.ts:181]: Comments to remove ?

File: twake/backend/node/src/services/user/services/users/service.ts [hunk 5]
```diff
@@ -74,4 +82,8 @@ export class UserService implements UsersServiceAPI {
   async get(pk: UserPrimaryKey): Promise<User> {
     return await this.repository.findOne(pk);
   }
+
+  async getUserCompanies(pk: UserPrimaryKey): Promise<ListResult<CompanyUser>> {
+    return await this.companyUserRepository.find({ user_id: pk.id });
```
[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/services/users/service.ts:87]: I thought the default is set to unlimited

File: twake/backend/node/src/services/user/web/controller.ts [hunk 6]
```diff
@@ -6,40 +6,200 @@ import { ResourceListResponse } from "../../../services/types";
...
+      ResourceListResponse<UserObject>,
+      ResourceDeleteResponse
+    > {
+  constructor(protected service: UsersServiceAPI, protected companyService: CompaniesServiceAPI) {
+    assert(service, "service is not inited");
+    assert(companyService, "companyService is not inited");
+    this.companiesCache = {};
+  }
+
+  companiesCache = {} as { [key: string]: Company };
```
[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/web/controller.ts:41]: It's cache for an instance just to reduce requests for the same company for two users. It invalidates on destruction.

obj → map: ok

File: twake/backend/node/src/services/user/web/controller.ts [hunk 7]
```diff
@@ -6,40 +6,200 @@ import { ResourceListResponse } from "../../../services/types";
...
+
+  private async getCompany(companyId: string): Promise<Company> {
+    const company: Company =
+      this.companiesCache[companyId] || (await this.companyService.getCompany({ id: companyId }));
+
+    this.companiesCache[companyId] = company;
+    return company;
+  }
+
+  private async getUserCompaniesOut(userId: string): Promise<UserCompanyObject[]> {
```
[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/web/controller.ts:51]: Hm… i've renamed it before push… that's odd… I have different name and «already up to date» for git… 
gonna check it

File: twake/backend/node/src/services/user/web/controller.ts [hunk 8]
```diff
@@ -6,40 +6,200 @@ import { ResourceListResponse } from "../../../services/types";
...
+
+    if (includeCompanies) {
+      resUser = {
+        ...resUser,
+        preference: {
+          locale: user.language,
+          timezone: user.timezone,
+        },
+
+        companies: await this.getUserCompaniesOut(user.id),
```
[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/web/controller.ts:95]: Don't get it … what do you mean by that? What's your proposition?

File: twake/backend/node/src/services/user/web/controller.ts [hunk 9]
```diff
@@ -6,40 +6,200 @@ import { ResourceListResponse } from "../../../services/types";
...
+export class UsersCrudController
+  implements
+    CrudController<
+      ResourceGetResponse<UserObject>,
+      ResourceCreateResponse<User>,
+      ResourceListResponse<UserObject>,
+      ResourceDeleteResponse
+    > {
+  constructor(protected service: UsersServiceAPI, protected companyService: CompaniesServiceAPI) {
+    assert(service, "service is not inited");
```
[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/web/controller.ts:36]: right

File: twake/backend/node/src/services/user/services/users/service.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/services/users/service.ts:87]: It's 100 for now… a situation in which a user in more than 100 companies is not really the case, but I got you — I'll add pagination opt

File: twake/backend/node/src/services/user/web/controller.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (RomaricMourgues) [reply twake/backend/node/src/services/user/web/controller.ts]: @chamerling would prefer to declare a variable first with the await and then put the variable into the resUser object:
```
const companies = await this.getUserCompaniesOut(user.id);
...
resUser = {
  ...
  companies,
  ...
}
```

[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/web/controller.ts]: done with variable

**Comment by RomaricMourgues** [comment]
Congratulation first PR on Twake :p 👏

File: twake/backend/node/src/services/user/web/controller.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (chamerling) [reply twake/backend/node/src/services/user/web/controller.ts]: @romanesko how do you deal with this when we have multiple NodeJS instances and a load balancer? What happens when a company is deleted (true, this is not so frequent), or updated, etc? This may not cause problems here but can cause some in some situation.
This is why cache in distributed environment is complex, and in our case, should be managed at the ORM level.

[the remark this answers is not in the document]
Comment (RomaricMourgues) [reply twake/backend/node/src/services/user/web/controller.ts]: Cache in single node is still a good idea. But yes, it should be moved to the ORM itself good idea.

[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/web/controller.ts]: It's not a cache in common sense.
It leaves within only a one request, it is just a trick, to economize a bit database queries

look, I have async requests:
```js
const [currentUserCompanies, requestedUserCompanies] = await Promise.all(
      [context.user.id, request.params.id].map(id =>
        this.service.getUserCompanies({ id }).then(a => a.getEntities()),
      ),
 );
    
```    
the companies may be common for both users, in this case if i get a company for one user I don't have to do query db for another (cause it is should be the same).

[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/web/controller.ts]: Did I misunderstand something?

[the remark this answers is not in the document]
Comment (romanesko) [reply twake/backend/node/src/services/user/web/controller.ts]: Soo I think I got it now —  don't we create a new instance of controller for every request?
